README.md
Nosefart plays [NSF](http://vgmpf.com/Wiki/index.php?title=NSF) files.

This program has changed hands many many times over the years and each person mostly just changed what they needed atm. This means it's generally buggy and unreliable, but I'm trying my best to change that.

Because it's had such a long heritage, many many variants of nosefart exist. From a gtk standalone player, to a winamp plugin, and many more. The code has been borrowed by other projects, and it borrows code from yet other projects.
In any case, this repository contains only what's necessary for the linux standalone command-line tool, which was always the primary target. Other continuations of nosefart might still include more things, but are usually much more outdated.

## Usage

```
Usage: nosefart [OPTIONS] filename
Play an NSF (NES Sound Format) file.

Options:
	-h  	Help
	-v  	Version information

	-t x	Play track x (default: 1)
	-s x	Play at x times the normal speed
	-f x	Use x sampling rate (default: 44100)
	-B x	Use sample size of x bits (default: 8)
	-l x	Limit total playing time to x seconds (0 = unlimited)
	-r x	Limit total playing time to x frames (0 = unlimited)
	-b x	Skip the first x frames
	-a x	Calculate song length and play x repetitions (0 = intro only)
	-i	Just print file information and exit
	-x	Start with channel x disabled (-123456)
	-o x	Output WAV file(s) to directory x
```

Nosefart uses SDL2, so make sure you have that.

## Credits

- Matthew Conte wrote the sound core and the original Linux and DOS programs.
  He says that nosefart was "inspired by the MSP 0.50 source release by Sevy and Marp \[and\] ported by Neil Stevens."
- Matthew Strait updated the Linux command line program and added gnosefart.
- Benjamin Gerard contributed the code that calculates song length.
- K`rai (a.k.a HJ) has fixed some bugs and worked on the Winamp plugin.
- Raphael Assenat made Linux nosefart work on big endian machines.
- Misterhat extracted core nosefart, ported it to SDL2 & added wav dumping capability.
- Riedler did lots of things, mostly fixing bugs and cleanup (see git log)
- Some other people are probably forgotten. :-(

CHANGES
3.3 (16 Aug 2023)

- fixed wav headers
- removed unused stereo parameters
- dumping now only does auto calc if specified

3.2 (15 Aug 2023)

- various cleanups
- fixed an out-of-bounds access
- removed lots of dead code from nsfinfo
- rewrote much of the time calculation code
- dumping now respects repetition parameter
- added new readme

3.1 (7 Aug 2023)

- now, only one track is dumped when it's selected
- output track lengths in file info
- fixed the last track being omitted when dumping (off-by-one caused by track indices starting at 1)
- minor changes to dev environment
- fully removed support for lots of deprecated systems and sound cards
- moved to C99 and cleaned up accordingly

3.0 (4 Aug 2021)

- removed everything outside core nosefart (quote: "legacy cruft")
- replaced OSS with SDL
- removed dos support
- removed threading
- added option to dump wav files
- added support for clang
- various smaller fixes

2.9-mls - 2013-05-10

GCC now requires that -lm comes after the object name.  Trivial change
to Makefile to allow compilation.  Thanks to Gall for reporting this.

2.8-mls

GNU Make 3.82 broke the Makefile so that it no longer made the necessary
build directories. Put in a simple hack to make it keep working.

2.7-mls

Added 16-bit fixes and -B switch to Linux nosefart by Chris King.

Improved gnosefart error message given when nosefart isn't found.

2.6-mls

Updated gnosefart to 1.4.

2.5-mls 

Added some lines to src/machine/nsf.c so that it's possible for programs
that use it to play multiple songs at once.  (This doesn't affect the
behavior of the Linux programs in this package at all. They still just
play one thing at a time.)

gnosefart updated to 1.3

2.4-mls (13 Sept 2005)

Updated nosefart code for gcc-4.0.

gnosefart updated to 1.2.

2.3-mls (12 Aug 2004)

Updated winamp plugin, thanks to K`rai.

2.2-mls (1 May 2004)

Added README.cl-amp (taken from Eli's page).

Added a fflush to nosefart for the benifit of gnosefart.

2.1-mls (23 April 2004)

gnosefart updated to 1.0.

2.0-mls (16 April 2004)

Added gnosefart 0.9, a GTK frontend for nosefart.

Decided to be more confident about my version numbering.

Nosefart now returns 0 if nothing went wrong.

1.92k-mls (4 April 2004)

Fixed a bug in the UI that made it not refresh properly.  Improved spec file.

1.92j-mls (21 March 2004)

A few small tweaks to nsfinfo and the interface.  Added spec file
for building RPMs.

1.92i-mls

Fixed hiccup that sometimes occurs when time calculation finishes.  The fix
is to simply not have the parent process wait() for the child, which means
that the child hangs around as a zombie as long as nosefart is running. Oh
well.

Fixed a bug that caused multiple free()s of the same thing to hang the time
calculation on certain files (lifeforce.nsf).

1.92h-mls

Time calculation (for Linux command line) now is done by a child process
which communicates the answer back through shared memory.  (If this doesn't
work, it falls back on the old behavior.) This prevents a long wait before
the music starts.

Added options to disable channels from the start (-1, -2, etc).

Added output of progress (i.e.  "4/94 seconds, 207/5589 frames").

1.92g-mls

Abstracted away the time calculation stuff so that it should be easy to make
it go with versions other than Linux/FreeBSD command line.

1.92f-mls

Added automatic track length calculation to the linux command line
version (-a option). (Developers: Note that a lot of ugly hacking was
involved here.)

Added xmms plugin.

Thanks to benjamin gerard <ben@sashipa.com> for both of these.

1.92e-mls

Added options -i (print file information and exit) and -l (play for x seconds
and then exit).

Fixed unaligned 16-bit memory access in the 6502 emulator.  This allows
nosefart to run on platforms other than x86 and PPC.

1.92d-mls

Added a Makefile to make it possible to compile a library suitable for dcplaya 
input plugin. 

1.92c-mls

Merged in the CL-amp plugin for BeOS.  I don't know if it actually works 
yet, since I don't have BeOS, but hopefully someone will test it and get 
back to me.

Made it compile in FreeBSD.  I don't know if it actually works, tho,
since I only have a shell account.  Someone tell me, please...

1.92b-mls

First sourceforge release.  I've improved the Linux frontend.  Now you can 
see what channels are enabled, these channels persist over track changes 
and if you advance past the last track, it loops back to the first.


src/machine/nsf.h
/*
** Nofrendo (c) 1998-2000 Matthew Conte (matt@conte.com)
**
**
** This program is free software; you can redistribute it and/or
** modify it under the terms of version 2 of the GNU Library General 
** Public License as published by the Free Software Foundation.
**
** This program is distributed in the hope that it will be useful, 
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU 
** Library General Public License for more details.  To obtain a 
** copy of the GNU Library General Public License, write to the Free 
** Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
**
** Any permitted reproduction of these routines, in whole or in part,
** must bear this legend.
**
**
** nsf.h
**
** NSF loading/saving related defines / prototypes
** $Id: nsf.h,v 1.3 2003/05/01 22:34:20 benjihan Exp $
*/

#ifndef _NSF_H_
#define _NSF_H_

#include "osd.h"
#include "nes6502.h"
#include "nes_apu.h"

#define  NSF_MAGIC   "NESM\x1A"

#define  NSF_DEDICATED_PAL       0x01
#define  NSF_DUAL_PAL_NTSC       0x02

#define  EXT_SOUND_NONE          0x00
#define  EXT_SOUND_VRCVI         0x01
#define  EXT_SOUND_VRCVII        0x02
#define  EXT_SOUND_FDS           0x04
#define  EXT_SOUND_MMC5          0x08
#define  EXT_SOUND_NAMCO106      0x10
#define  EXT_SOUND_SUNSOFT_FME07 0x20
/* bits 6,7: future expansion */

#define  NSF_HEADER_SIZE         0x80

/* 60 Hertz refresh (NTSC) */
#define  NES_MASTER_CLOCK     21477272.7272
#define  NTSC_REFRESH         60
#define  NTSC_SUBCARRIER_DIV  12
#define  NTSC_SCANLINES       262

#define  NES_FRAME_CYCLES     ((NES_MASTER_CLOCK / NTSC_SUBCARRIER_DIV) / NTSC_REFRESH)
#define  NES_SCANLINE_CYCLES  (NES_FRAME_CYCLES / NTSC_SCANLINES)

/* filter levels */
enum
{
   NSF_FILTER_NONE,
   NSF_FILTER_LOWPASS,
   NSF_FILTER_WEIGHTED,
   NSF_FILTER_MAX, /* $$$ ben : add this one for range chacking */
};

typedef struct nsf_s
{
   /* NESM header */
   uint8  id[5]               ; /* NESM\x1A */
   uint8  version             ; /* spec version */
   uint8  num_songs           ; /* total num songs */
   uint8  start_song          ; /* first song */
   uint16 load_addr           ; /* loc to load code */
   uint16 init_addr           ; /* init call address */
   uint16 play_addr           ; /* play call address */
   uint8  song_name[32]       ; /* name of song */
   uint8  artist_name[32]     ; /* artist name */
   uint8  copyright[32]       ; /* copyright info */
   uint16 ntsc_speed          ; /* playback speed (if NTSC) */
   uint8  bankswitch_info[8]  ; /* initial code banking */
   uint16 pal_speed           ; /* playback speed (if PAL) */
   uint8  pal_ntsc_bits       ; /* NTSC/PAL determination bits */
   uint8  ext_sound_type      ; /* type of external sound gen. */
   uint8  reserved[4]         ; /* reserved */

   /* things that the NSF player needs */
   uint8  *data;              /* actual NSF data */
   uint8  *acc_data;          /* shadow NSF data for access control */
   uint32 length;             /* length of data */
   uint32 playback_rate;      /* current playback rate */
   uint8  current_song;       /* current song */
   boolean bankswitched;      /* is bankswitched? */

  /* $$$ ben : Playing time ... */
  uint32 cur_frame;
  uint32 cur_frame_end;
  uint32 * song_frames;

  /* $$$ ben : Last error string */
   const char * errstr;       

   /* CPU and APU contexts */
   nes6502_context *cpu;
   apu_t *apu;

   /* our main processing routine, calls all external mixing routines */
   void (*process)(void *buffer, int num_samples);
} nsf_t;

/* $$$ ben : Generic loader struct */
struct nsf_loader_t {
  /* Init and open. */
  int (*open)(struct nsf_loader_t * loader);

  /* Close and shutdown. */
  void (*close) (struct nsf_loader_t * loader);

  /* This function should return <0 on error, else the number of byte NOT read.
   * that way a simple 0 test tell us if read was complete.
   */
  int (*read) (struct nsf_loader_t * loader, void *data, int n);

  /* Get file length. */
  int (*length) (struct nsf_loader_t * loader);

  /* Skip n bytes. */
  int (*skip) (struct nsf_loader_t * loader,int n);

  /* Get filename (for debug). */
  const char * (*fname) (struct nsf_loader_t * loader);

};

/* Function prototypes */
extern int nsf_init(void);

extern nsf_t * nsf_load_extended(struct nsf_loader_t * loader);
extern nsf_t *nsf_load(const char *filename, void *source, int length);
extern void nsf_free(nsf_t **nsf_info);

extern int nsf_playtrack(nsf_t *nsf, int track, int sample_rate,
			 int sample_bits);
extern void nsf_frame(nsf_t *nsf);
extern int nsf_setchan(nsf_t *nsf, int chan, boolean enabled);
extern int nsf_setfilter(nsf_t *nsf, int filter_type);

#endif /* _NSF_H_ */

/*
** $Log: nsf.h,v $
** Revision 1.3  2003/05/01 22:34:20  benjihan
** New NSF plugin
**
** Revision 1.2  2003/04/09 14:50:32  ben
** Clean NSF api.
**
** Revision 1.1  2003/04/08 20:53:00  ben
** Adding more files...
**
** Revision 1.11  2000/07/04 04:59:24  matt
** removed DOS-specific stuff
**
** Revision 1.10  2000/07/03 02:19:36  matt
** dynamic address range handlers, cleaner and faster
**
** Revision 1.9  2000/06/23 03:27:58  matt
** cleaned up external sound inteface
**
** Revision 1.8  2000/06/20 04:04:37  matt
** moved external soundchip struct to apu module
**
** Revision 1.7  2000/06/20 00:05:45  matt
** changed to driver-based external sound generation
**
** Revision 1.6  2000/06/13 03:51:54  matt
** update API to take freq/sample data on nsf_playtrack
**
** Revision 1.5  2000/06/12 01:13:00  matt
** added CPU/APU as members of the nsf struct
**
** Revision 1.4  2000/06/09 15:12:26  matt
** initial revision
**
*/
